pkg/ebpf/c/tracee.bpf.c

```diff
@@ -6380,3 +6406,812 @@ int tc_ingress(struct __sk_buff *skb)
+int BPF_KPROBE(trace_sock_alloc_file)
+{
+    // runs every time a socket is created (entry)
+
+    event_data_t data = {};
+
+    if (!init_event_data(&data, ctx))
+        return 0;
+
+    if (!should_trace(&data))
+        return 0;
+
+    struct entry entry = {0};
+
+    // save args for retprobe
+    entry.args[0] = PT_REGS_PARM1(ctx); // struct socket *sock
+    entry.args[1] = PT_REGS_PARM2(ctx); // int flags
+    entry.args[2] = PT_REGS_PARM2(ctx); // char *dname
+
+    struct socket *sock = (void *) PT_REGS_PARM1(ctx);
+
+    if (!is_socket_supported(sock))
+        return 0;
+
+    // prepare for kretprobe using entrymap
+    u32 host_tid = data.context.task.host_tid;
+    bpf_map_update_elem(&entrymap, &host_tid, &entry, BPF_ANY);
+
+    return 0;
+}
+
+SEC("kretprobe/sock_alloc_file")
+int BPF_KRETPROBE(trace_ret_sock_alloc_file)
+{
+    // runs every time a socket is created (return)
+
+    event_data_t data = {};
+    if (!init_event_data(&data, ctx))
+        return 0;
+
+    // pick from entry from entrymap
+    u32 host_tid = data.context.task.host_tid;
+    struct entry *entry = bpf_map_lookup_elem(&entrymap, &host_tid);
+    if (!entry) // no entry == no tracing
+        return 0;
+
+    // pick args from entry point's entry
+    struct socket *sock = (void *) entry->args[0];
+    // int flags = entry->args[1];
+    // char *dname = (void *) entry->args[2];
+    struct file *sock_file = (void *) PT_REGS_RC(ctx);
+
+    // cleanup entrymap
+    bpf_map_delete_elem(&entrymap, &host_tid);
+
+    if (!sock_file)
+        return 0; // socket() failed ?
+
+    u64 inode = BPF_READ(sock_file, f_inode, i_ino);
+    if (inode == 0)
+        return 0;
+
+    // save context to further create an event when no context exists
+    net_task_context_t netctx = {0};
+    set_net_task_context(&data, &netctx);
+
+    // update inodemap correlating inode <=> task
+    bpf_map_update_elem(&inodemap, &inode, &netctx, BPF_ANY);
+
+    return 0;
+}
+
+//
+// Socket Ingress/Egress eBPF program loader (right before and right after eBPF)
+//
+
+SEC("kprobe/__cgroup_bpf_run_filter_skb")
+int BPF_KPROBE(cgroup_bpf_run_filter_skb)
+{
+    // runs BEFORE the CGROUP/SKB eBPF program
+
+    event_data_t data = {};
+
+    if (!init_event_data(&data, ctx))
+        return 0;
+
+    struct sock *sk = (void *) PT_REGS_PARM1(ctx);
+    struct sk_buff *skb = (void *) PT_REGS_PARM2(ctx);
+    int type = PT_REGS_PARM3(ctx);
+
+    switch (type) {
+        case BPF_CGROUP_INET_INGRESS:
+        case BPF_CGROUP_INET_EGRESS:
+            break;
+        default:
+            return 0; // wrong attachment type, return fast
+    }
+
+    // obtain socket inode
+    u64 inode = BPF_READ(sk, sk_socket, file, f_inode, i_ino);
+    if (inode == 0)
+        return 0;
+
+    // save args for kretprobe
+    struct entry entry = {0};
+    entry.args[0] = PT_REGS_PARM1(ctx); // struct sock *sk
+    entry.args[1] = PT_REGS_PARM2(ctx); // struct sk_buff *skb
```

Comment:
Another possible optimization.

```suggestion
    int type = PT_REGS_PARM3(ctx);

    switch (type) {
        case BPF_CGROUP_INET_INGRESS:
        case BPF_CGROUP_INET_EGRESS:
            break;
        default:
            return 0; // wrong attachment type, return fast
    }
    
    struct sock *sk = (void *) PT_REGS_PARM1(ctx);

    // obtain socket inode
    u64 inode = BPF_READ(sk, sk_socket, file, f_inode, i_ino);
    if (inode == 0)
        return 0;

    struct sk_buff *skb = (void *) PT_REGS_PARM2(ctx);

    // save args for kretprobe
    struct entry entry = {0};
    entry.args[0] = (unsigned long) sk;  // struct sock *sk
    entry.args[1] = (unsigned long) skb; // struct sk_buff *skb
```